### PR TITLE
Implement `provablyDisjoint` for refined types and type aliases

### LINKED ISSUE #15312: Type matching on structural types does not work (nor gives an error)
## Compiler version

3.1.2

## Minimized code

```Scala
type F[t] =
  t match
    case {type A = Float} => Int
    case {type A = Int} => String

val a: F[{type A = Float}] = 10
val b: F[{type A = Int}] = "asd" // Found:("asd" : String) Required: F[Object{A = Int}]
```

## Output

```scala
Found:("asd" : String) Required: F[Object{A = Int}]
```

## Expectation

I expect the compiler to either do the right match (accepting string value for `val b`) or give some diagnostic about the limitation of match types not playing with structural patterns

### PR DESCRIPTION
**Pull request by odersky**
Fixes #15312

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -2659,6 +2659,10 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
         fullyInstantiated(tp2) && !tp1.classSymbols.exists(_.derivesFrom(tp2.symbol))
       case (tp1: TypeRef, tp2: TermRef) if isEnumValue(tp2) =>
         fullyInstantiated(tp1) && !tp2.classSymbols.exists(_.derivesFrom(tp1.symbol))
+      case (tp1: RefinedType, tp2: RefinedType) if tp1.refinedName == tp2.refinedName =>
+        provablyDisjoint(tp1.parent, tp2.parent) || provablyDisjoint(tp1.refinedInfo, tp2.refinedInfo)
+      case (tp1: TypeAlias, tp2: TypeAlias) =>
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:2664]: The overall expression matches on `(tp1.dealias, tp2.dealias)`, does it make that case unreachable?

**Review by OlivierBlanvillain — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:2664]: No, dealias does not unwind refined types.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:2664]: I think there was a miscommunication: this (TypeAlias, TypeAlias) case seems redundant indeed.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:2664]: Wait, what?  Dealias doesn't dealias TypeAlias's?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:2664]: No, it dealiases TypeRefs whose inofo is a TypeAlias. A TypeAlias is a subtype of TypeBounds. It is information _about_ a type, not a type itself.

**Pull request merged**